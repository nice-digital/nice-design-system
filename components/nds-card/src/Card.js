import React, { cloneElement } from "react";
import classnames from "classnames";
import PropTypes from "prop-types";
import "../scss/card.scss";

const CardHeader = props => {
	const { headingText, headingElementType: HeadingTag = "p", link } = props;
	let linkProps = {};
	if (link) {
		const { elementType: LinkTag = "a", destination } = link;
		if (LinkTag === "a") {
			linkProps.href = destination;
		} else {
			linkProps.to = destination;
		}
		return (
			<HeadingTag className="card__heading">
				<LinkTag {...linkProps}>{headingText}</LinkTag>
			</HeadingTag>
		);
	} else {
		return <HeadingTag className="card__heading">{headingText}</HeadingTag>;
	}
};

export const Card = props => {
	const {
		metadata,
		headingText,
		headingElementType,
		link,
		image,
		children,
		elementType: ContainerType = "article",
		...rest
	} = props;
	const headerProps = { headingText, headingElementType, link };
	return (
		<ContainerType className="card" {...rest}>
			{image && <div className="card__image">{image}</div>}
			<div className="card__text">
				<header className="card__header">
					<CardHeader {...headerProps} />
				</header>
				{children && <div className="card__summary">{children}</div>}
				{metadata && metadata.length && (
					<dl className="card__metadata">
						{metadata.map((item, idx) => {
							if (!item.value) return null;
							return (
								<div key={`item${idx}`} className="card__metadatum">
									{item.label && (
										<dt className="visually-hidden">{item.label}</dt>
									)}
									<dd>{item.value}</dd>
								</div>
							);
						})}
					</dl>
				)}
			</div>
		</ContainerType>
	);
};

Card.propTypes = {
	children: PropTypes.node,
	elementType: PropTypes.elementType,
	headingElementType: PropTypes.elementType,
	headingText: PropTypes.node.isRequired,
	image: PropTypes.node,
	link: PropTypes.shape({
		destination: PropTypes.node,
		elementType: PropTypes.elementType
	}),
	metadata: PropTypes.arrayOf(
		PropTypes.shape({
			label: PropTypes.node,
			value: PropTypes.node.isRequired
		})
	)
};
