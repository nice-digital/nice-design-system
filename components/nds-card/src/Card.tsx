import React from "react";
import classNames from "classnames";
import "../scss/card.scss";

export interface CardHeadingLinkProps {
	[prop: string]: unknown;
	destination?: string;
	elementType?: React.ElementType;
	method?: string;
}

interface CardHeaderProps {
	headingElementType?: React.ElementType;
	headingText: React.ReactNode;
	link: CardHeadingLinkProps;
}

const CardHeader = (props: CardHeaderProps) => {
	const { headingText, headingElementType: HeadingTag = "p", link } = props;

	let linkProps = {};

	if (link) {
		const { elementType: ElementType = "a", destination, method } = link;

		linkProps = {
			[method || (ElementType === "a" && "href") || "to"]: destination
		};

		return (
			<HeadingTag className="card__heading">
				<ElementType {...linkProps}>{headingText}</ElementType>
			</HeadingTag>
		);
	} else {
		return <HeadingTag className="card__heading">{headingText}</HeadingTag>;
	}
};

export interface CardBodyProps {
	[prop: string]: unknown;
	children?: React.ReactNode;
	summary?: React.ReactNode;
	headingElementType?: React.ElementType;
	headingText: React.ReactNode;
	link?: CardHeadingLinkProps;
	metadata?: CardMetaDataProps[];
}

interface CardMetaDataProps {
	[prop: string]: unknown;
	label?: string;
	value: React.ReactNode;
	visibleLabel?: boolean;
}

const CardBody = (props: CardBodyProps) => {
	const { metadata, headingText, headingElementType, link, summary, children } =
		props;
	const headerProps = {
		headingText,
		headingElementType: headingElementType as React.ElementType,
		link: link as CardHeadingLinkProps
	};
	return (
		<>
			<header className="card__header">
				<CardHeader {...headerProps} />
			</header>
			{summary && <p className="card__summary">{summary}</p>}

			{children && children}

			{metadata && metadata.length && (
				<dl className="card__metadata">
					{metadata.map((item, idx) => {
						if (!item.value) return null;
						return (
							<div key={`item${idx}`} className="card__metadatum">
								{item.label && (
									<dt
										className={classNames({
											"visually-hidden": !item.visibleLabel
										})}
									>
										{item.label}
									</dt>
								)}
								<dd>{item.value}</dd>
							</div>
						);
					})}
				</dl>
			)}
		</>
	);
};

export interface CardProps {
	[prop: string]: unknown;
	children?: React.ReactNode;
	summary?: React.ReactNode;
	elementType?: React.ElementType;
	headingElementType?: React.ElementType;
	headingText: React.ReactNode;
	image?: React.ReactNode;
	link?: CardHeadingLinkProps;
	metadata?: Array<CardMetaDataProps>;
	className?: string;
}

export const Card = (props: CardProps) => {
	const {
		className,
		metadata,
		headingText,
		headingElementType,
		link,
		image,
		summary,
		elementType: ContainerType = "article",
		children,
		...rest
	} = props;

	const cardBodyProps = {
		metadata,
		headingText,
		headingElementType,
		link,
		summary,
		children
	};

	const classes = classNames(["card", `${className}`]);

	return (
		<ContainerType className={classes} {...rest}>
			{image ? (
				<>
					<div className="card__image">{image}</div>
					<div className="card__text">
						<CardBody {...cardBodyProps} />
					</div>
				</>
			) : (
				<CardBody {...cardBodyProps} />
			)}
		</ContainerType>
	);
};
