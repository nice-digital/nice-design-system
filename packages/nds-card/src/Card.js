import React from "react";
import PropTypes from "prop-types";
import "../scss/card.scss";

const CardHeader = props => {
	const { headingText, elementType: HeadingTag = "p", link } = props;
	let linkProps = {};
	if (link) {
		const { elementType: LinkTag = "a", destination } = link;
		if (LinkTag === "a") {
			linkProps.href = destination;
		} else {
			linkProps.to = destination;
		}
		return (
			<HeadingTag className="card__heading">
				<LinkTag {...linkProps}>{headingText}</LinkTag>
			</HeadingTag>
		);
	} else {
		return <HeadingTag className="card__heading">{headingText}</HeadingTag>;
	}
};

export const Card = props => {
	const { metadata } = props;
	return (
		<article className="card">
			<header className="card__header">
				<CardHeader {...props} />
			</header>
			{metadata && metadata.length && (
				<dl className="card__metadata">
					{metadata.map((item, idx) => {
						if (!item.value) return null;
						return (
							<div key={`item${idx}`} className="card__metadatum">
								{item.label && (
									<dt className="visually-hidden">{item.label}</dt>
								)}
								<dd>{item.value}</dd>
							</div>
						);
					})}
				</dl>
			)}
		</article>
	);
};

Card.propTypes = {
	headingText: PropTypes.node.isRequired,
	elementType: PropTypes.elementType,
	link: PropTypes.shape({
		destination: PropTypes.node,
		elementType: PropTypes.elementType
	}),
	metadata: PropTypes.arrayOf(
		PropTypes.shape({
			label: PropTypes.node,
			value: PropTypes.node.isRequired
		})
	)
};
