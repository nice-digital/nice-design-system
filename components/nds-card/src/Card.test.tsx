import React from "react";
import { render } from "@testing-library/react";

import { Card, CardProps, CardHeadingLinkProps } from "./Card";

const headingProps: CardHeadingLinkProps = {
	headingText: "Card heading text",
	headingElementType: "h1",
	link: {
		destination: "/about",
		elementType: "a"
	}
};
const metadataProps = [
	{
		value: "john.smith@email.com",
		label: "Email address"
	},
	{
		value: "sarah.smith@email.com",
		label: "Email address"
	}
];

const props: CardProps = {
	headingText: "Card heading text",
	headingElementType: "h1",
	link: {
		destination: "/about",
		elementType: "a"
	},
	metadata: metadataProps
};

describe("Card", () => {
	it("should match snapshot with some default attributes", () => {
		const wrapper = render(
			<>
				<Card {...props} summary="I'm just a string" />
				<Card
					{...props}
					image={<img src="test.png" alt="Cute dog" />}
					summary={
						<span>
							how <b>bold</b> of you
						</span>
					}
				/>
			</>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("should match snapshot for the callout variant", () => {
		const wrapper = render(
			<>
				<Card {...props} summary="I'm just a string" callout={true} />
				<Card
					{...props}
					image={<img src="test.png" alt="Cute dog" />}
					summary={
						<span>
							how <b>bold</b> of you
						</span>
					}
					callout={true}
				/>
			</>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it("should pass extra props to the containing element", () => {
		const localProps = Object.assign({}, props, {
			"data-tracker": "my-tracker"
		});
		const { container } = render(<Card {...localProps} />);
		expect(container.querySelector(".card")?.getAttribute("data-tracker")).toBe(
			"my-tracker"
		);
	});

	it("should create a card with the containing element type as supplied", () => {
		const wrapper = render(<Card headingText="My Heading" elementType="li" />);
		expect(wrapper.getByRole("listitem")).toBeInTheDocument();
	});

	it("should render each metadata item with a value", () => {
		const { container } = render(<Card {...props} />);
		expect(container.querySelectorAll(".card__metadatum").length).toBe(2);
	});

	it("should not render a metadata item without a value", () => {
		const localMetadataProps = [...metadataProps];
		localMetadataProps[0].value = "";
		const { container } = render(
			<Card {...props} metadata={localMetadataProps} />
		);
		expect(container.querySelectorAll(".card__metadatum").length).toBe(1);
	});

	it("should render an href attribute if the elementType is an anchor", () => {
		const wrapper = render(<Card {...props} />);
		const anchor = wrapper.getByRole("link");
		expect(anchor.getAttribute("href")).toBe("/about");
	});

	it("should pass a custom navigaiton method if supplied", () => {
		const localProps = Object.assign({}, props, {
			link: { method: "pigeon", destination: "/about" }
		});
		const { container } = render(<Card {...localProps} />);
		const anchor = container.querySelector("a");
		expect(anchor?.getAttribute("pigeon")).toBe("/about");
	});

	it("should not hide label if a metadatum has a visibleLabel property", () => {
		const { container } = render(
			<Card
				headingText="Hello"
				metadata={[
					{
						label: "Email address",
						value: "test@test.com"
					},
					{
						label: "Published on",
						value: "27 May 2012",
						visibleLabel: true
					},
					{
						label: "Reviewed on",
						value: "27 May 2022",
						visibleLabel: false
					}
				]}
			/>
		);
		expect(container.querySelectorAll("dt.visually-hidden").length).toBe(2);
		expect(container.querySelectorAll("dt").length).toBe(3);
	});

	it("should render a wrapped child element", () => {
		const { container } = render(
			<>
				<Card headingText="Children rendering">
					<b id="child">Render child</b>
				</Card>
			</>
		);
		const childElement = container.querySelector("#child");
		expect(childElement?.textContent).toBe("Render child");
	});
});
