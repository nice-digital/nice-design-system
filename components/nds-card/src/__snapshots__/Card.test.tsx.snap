// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot with some default attributes 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="card"
      >
        <header
          class="card__header"
        >
          <h1
            class="card__heading"
          >
            <a
              href="/about"
            >
              Card heading text
            </a>
          </h1>
        </header>
        <p
          class="card__summary"
        >
          I'm just a string
        </p>
        <dl
          class="card__metadata"
        >
          <div
            class="card__metadatum"
          >
            <dt
              class="visually-hidden"
            >
              Email address
            </dt>
            <dd>
              john.smith@email.com
            </dd>
          </div>
          <div
            class="card__metadatum"
          >
            <dt
              class="visually-hidden"
            >
              Email address
            </dt>
            <dd>
              sarah.smith@email.com
            </dd>
          </div>
        </dl>
      </article>
      <article
        class="card"
      >
        <div
          class="card__image"
        >
          <img
            alt="Cute dog"
            src="test.png"
          />
        </div>
        <div
          class="card__text"
        >
          <header
            class="card__header"
          >
            <h1
              class="card__heading"
            >
              <a
                href="/about"
              >
                Card heading text
              </a>
            </h1>
          </header>
          <p
            class="card__summary"
          >
            <span>
              how 
              <b>
                bold
              </b>
               of you
            </span>
          </p>
          <dl
            class="card__metadata"
          >
            <div
              class="card__metadatum"
            >
              <dt
                class="visually-hidden"
              >
                Email address
              </dt>
              <dd>
                john.smith@email.com
              </dd>
            </div>
            <div
              class="card__metadatum"
            >
              <dt
                class="visually-hidden"
              >
                Email address
              </dt>
              <dd>
                sarah.smith@email.com
              </dd>
            </div>
          </dl>
        </div>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="card"
    >
      <header
        class="card__header"
      >
        <h1
          class="card__heading"
        >
          <a
            href="/about"
          >
            Card heading text
          </a>
        </h1>
      </header>
      <p
        class="card__summary"
      >
        I'm just a string
      </p>
      <dl
        class="card__metadata"
      >
        <div
          class="card__metadatum"
        >
          <dt
            class="visually-hidden"
          >
            Email address
          </dt>
          <dd>
            john.smith@email.com
          </dd>
        </div>
        <div
          class="card__metadatum"
        >
          <dt
            class="visually-hidden"
          >
            Email address
          </dt>
          <dd>
            sarah.smith@email.com
          </dd>
        </div>
      </dl>
    </article>
    <article
      class="card"
    >
      <div
        class="card__image"
      >
        <img
          alt="Cute dog"
          src="test.png"
        />
      </div>
      <div
        class="card__text"
      >
        <header
          class="card__header"
        >
          <h1
            class="card__heading"
          >
            <a
              href="/about"
            >
              Card heading text
            </a>
          </h1>
        </header>
        <p
          class="card__summary"
        >
          <span>
            how 
            <b>
              bold
            </b>
             of you
          </span>
        </p>
        <dl
          class="card__metadata"
        >
          <div
            class="card__metadatum"
          >
            <dt
              class="visually-hidden"
            >
              Email address
            </dt>
            <dd>
              john.smith@email.com
            </dd>
          </div>
          <div
            class="card__metadatum"
          >
            <dt
              class="visually-hidden"
            >
              Email address
            </dt>
            <dd>
              sarah.smith@email.com
            </dd>
          </div>
        </dl>
      </div>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
